### PR TITLE
refactor: 1306 fix save changes

### PR DESCRIPTION
**Pull request by jonasdmoeller**
## Description

Refactored Charge Repository to search local context before DB lookup. This enables us to identify if a charge is already created/added in local context (but not committed) during charge update flow.

The temporary fix to save unit of work after charge create is now removed.

### CODE REVIEW

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Persistence/Repositories/ChargeRepository.cs [hunk 1]
```diff
@@ -60,24 +65,48 @@ public async Task AddAsync(Charge charge)
...
+                    c.Type == chargeIdentifier.ChargeType)
+                .ConfigureAwait(false);
+        }
+
+        private async Task<Charge> SingleFromDbAsync(Guid id)
+        {
+            return await _chargesDatabaseContext.Charges
+                .SingleAsync(c =>
+                    c.Id == id)
+                .ConfigureAwait(false);
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Persistence/Repositories/ChargeRepository.cs:99]: Should be covered by a test in `ChargeRepositoryTests`

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Persistence/Repositories/MarketParticipantRepository.cs [hunk 2]
```diff
@@ -31,18 +31,19 @@ public class MarketParticipantRepository : IMarketParticipantRepository
...
         }
 
-        public async Task<MarketParticipant?> GetOrNullAsync(string marketParticipantId)
+        public async Task<MarketParticipant?> SingleOrNullAsync(string marketParticipantId)
         {
             return await _chargesDatabaseContext
                 .MarketParticipants
-                .SingleOrDefaultAsync(mp => mp.MarketParticipantId == marketParticipantId).ConfigureAwait(false);
+                .SingleOrDefaultAsync(mp => mp.MarketParticipantId == marketParticipantId)
+                .ConfigureAwait(false);
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Persistence/Repositories/MarketParticipantRepository.cs:46]: Like for `ChargeRepository`, let's also test the new parts of `MarketParticipantRepository`.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/IntegrationTests/Repositories/ChargeRepositoryTests.cs [hunk 3]
```diff
@@ -90,66 +91,81 @@ public async Task AddAsync_WhenChargeIsNull_ThrowsArgumentNullException(ChargeRe
...
+            var key = new ChargeIdentifier(
+                charge.SenderProvidedChargeId,
+                charge.OwnerId,
+                charge.Type);
+            var actual = await sut.SingleAsync(key);
 
             // Assert
-            actual.Should().NotBeNull();
-            actual.Should().BeEquivalentTo(charge);
+            Assert.NotNull(actual);
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/IntegrationTests/Repositories/ChargeRepositoryTests.cs:113]:
```suggestion
            actual.Should().NotBeNull();
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.MessageHub/Models/AvailableChargeLinksData/AvailableChargeLinksDataFactory.cs [hunk 4]
```diff
@@ -64,10 +64,16 @@ public override async Task<IReadOnlyList<AvailableChargeLinksData>> CreateAsync(
             // not considered part of the Create Metering Point orchestration.
             // We select the first as all bundled messages will have the same recipient
             var recipient = await _marketParticipantRepository
-                .GetGridAccessProviderAsync(operation.MeteringPointId).ConfigureAwait(false);
+                .GetGridAccessProviderAsync(operation.MeteringPointId)
+                .ConfigureAwait(false);
 
-            var chargeIdentifier = new ChargeIdentifier(operation.SenderProvidedChargeId, operation.ChargeOwner, operation.ChargeType);
-            var charge = await _chargeRepository.GetAsync(chargeIdentifier).ConfigureAwait(false);
+            // TODO INDEN PR: Spørg Peter eller Irene om recipient altid er lig owner
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.MessageHub/Models/AvailableChargeLinksData/AvailableChargeLinksDataFactory.cs:70]: @prtandrup @IreneDataHub

**Review by HenrikSommer — COMMENTED** [review]
Left a few comments, primarily minor suggestions

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/IChargeRepository.cs [hunk 5]
```diff
@@ -25,12 +25,10 @@ public interface IChargeRepository
     {
         Task AddAsync(Charge charge);
 
-        Task<Charge> GetAsync(ChargeIdentifier chargeIdentifier);
+        Task<Charge> SingleAsync(ChargeIdentifier chargeIdentifier);
 
-        Task<Charge> GetAsync(Guid id);
+        Task<IReadOnlyCollection<Charge>> SingleAsync(IReadOnlyCollection<Guid> ids);
```
[the remark this answers is not in the document]
Comment (Mech0z) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/IChargeRepository.cs:30]: renamed to GetByIds

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Persistence/Repositories/ChargeRepository.cs [hunk 6]
```diff
@@ -31,27 +32,31 @@ public ChargeRepository(IChargesDatabaseContext chargesDatabaseContext)
...
+            return new ReadOnlyCollection<Charge>(charges);
         }
 
-        public async Task<Charge?> GetOrNullAsync(ChargeIdentifier chargeIdentifier)
+        public async Task<Charge?> SingleOrNullAsync(ChargeIdentifier chargeIdentifier)
         {
-            return await GetChargeQueryable(chargeIdentifier).SingleOrDefaultAsync().ConfigureAwait(false);
+            var charge = SingleOrDefaultLocal(chargeIdentifier) ??
+                         await SingleOrNullFromDbAsync(chargeIdentifier).ConfigureAwait(false);
+            return charge;
```
[the remark this answers is not in the document]
Comment (Mech0z) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Persistence/Repositories/ChargeRepository.cs:63]: done

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/ActorProvider.cs [hunk 7]
```diff
@@ -32,7 +32,7 @@ public ActorProvider(IMarketParticipantRepository marketParticipantRepository)
 
         public async Task<Actor> GetActorAsync(Guid actorId)
         {
-            var mp = await _marketParticipantRepository.GetOrNullAsync(actorId).ConfigureAwait(false);
+            var mp = await _marketParticipantRepository.SingleOrNullAsync(actorId).ConfigureAwait(false);
```
[the remark this answers is not in the document]
Comment (Mech0z) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/ActorProvider.cs:35]: Added ActorProvider tests

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Persistence/Repositories/ChargeRepository.cs [hunk 8]
```diff
@@ -31,27 +32,35 @@ public ChargeRepository(IChargesDatabaseContext chargesDatabaseContext)
...
+                         await SingleFromDbAsync(chargeIdentifier).ConfigureAwait(false);
+            return charge;
         }
 
-        public Task<Charge> GetAsync(Guid id)
+        public async Task<IReadOnlyCollection<Charge>> GetByIdsAsync(IReadOnlyCollection<Guid> ids)
         {
-            return GetChargesAsQueryable().SingleAsync(x => x.Id == id);
-        }
+            var charges = new List<Charge>();
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Persistence/Repositories/ChargeRepository.cs:44]: The `charges` list is populated, but unused. Did you mean to return this list in L#56?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/IntegrationTests/Repositories/ChargeRepositoryTests.cs [hunk 9]
```diff
@@ -90,66 +91,100 @@ public async Task AddAsync_WhenChargeIsNull_ThrowsArgumentNullException(ChargeRe
...
 
             // Assert
-            actual.Should().NotBeEmpty();
+            actual.First().OwnerId.Should().Be(firstCharge.OwnerId);
+            actual.Last().OwnerId.Should().Be(secondCharge.OwnerId);
+        }
+
+        [Theory]
+        [InlineAutoMoqData]
+        public async Task GetByIdsAsync_VerifyLocalContextItemsAreAlsoFetched(Charge notSavedCharge)
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/IntegrationTests/Repositories/ChargeRepositoryTests.cs:170]: The test doesn't verify that items are loaded from the database.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Persistence/Repositories/ChargeRepository.cs (hunk 8 above)
Comment (Mech0z) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Persistence/Repositories/ChargeRepository.cs:44]: fixed

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/IntegrationTests/Repositories/ChargeRepositoryTests.cs (hunk 9 above)
Comment (Mech0z) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/IntegrationTests/Repositories/ChargeRepositoryTests.cs:170]: added https://github.com/Energinet-DataHub/geh-charges/pull/1362/commits/b8b11aa3461efc82d2154e2027d0e0f836ed1a00

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Persistence/Repositories/ChargeRepository.cs [hunk 10]
```diff
@@ -60,24 +69,48 @@ public async Task AddAsync(Charge charge)
...
+        private Charge? SingleOrDefaultLocal(ChargeIdentifier chargeIdentifier)
+        {
+            return _chargesDatabaseContext.Charges
+                .Local.SingleOrDefault(c =>
+                    c.SenderProvidedChargeId == chargeIdentifier.SenderProvidedChargeId &&
+                    c.OwnerId == chargeIdentifier.Owner &&
+                    c.Type == chargeIdentifier.ChargeType);
+        }
+
+        private Charge? SingleOrDefaultLocal(Guid id)
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Persistence/Repositories/ChargeRepository.cs:81]: Doesn't seem to be used. Remove?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Persistence/Repositories/ChargeRepository.cs [hunk 11]
```diff
@@ -60,24 +69,48 @@ public async Task AddAsync(Charge charge)
...
-            return query;
+            return await _chargesDatabaseContext.Charges
+                .SingleAsync(c =>
+                    c.SenderProvidedChargeId == chargeIdentifier.SenderProvidedChargeId &&
+                    c.OwnerId == chargeIdentifier.Owner &&
+                    c.Type == chargeIdentifier.ChargeType)
+                .ConfigureAwait(false);
+        }
+
+        private async Task<Charge> SingleFromDbAsync(Guid id)
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Persistence/Repositories/ChargeRepository.cs:98]: Doesnt seem to be used?

**Review by HenrikSommer — APPROVED** [review]

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Persistence/Repositories/ChargeRepository.cs (hunk 11 above)
Comment (Mech0z) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Persistence/Repositories/ChargeRepository.cs]: done

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.MessageHub/Models/AvailableChargeLinksData/AvailableChargeLinksDataFactory.cs (hunk 4 above)
Comment (prtandrup) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.MessageHub/Models/AvailableChargeLinksData/AvailableChargeLinksDataFactory.cs]: Mads and I just talked about it. Code looks fine. The recipient and owner are not identical.

**Pull request merged**